specs/disk_details_for_hardware_inventory.adoc

```diff
@@ -0,0 +1,178 @@
+= Specification for adding disk details in node hardware inventory
+
+Add disk details also as a part of node hardware inventory
+
+== Problem description
+
+Hardware inventory details are used to show exact status of the particular
+node. Currently disk details are missing in node hardware inventory. Add disk
+details in node hardware inventory and disk details should contain disk_id,
+device_name, fs_type, fs_uuid, model, mount_point, parent, size, name, disk_type,
+used, ssd, vendor, serial_no, firmware_version.
+
+== Use Cases
+
+Tendrl should show the current status of disk.
+
+== Proposed change
+
+* "hwinfo" and "lsblk" to find the disk details.
+* lsblk is not sufficient to find all disk details, So we have to use hwinfo also.
+* From node_agent add a new function in pull_hardware_inventory to find disk
+  details and update those details into node_inventory list.
+* Create three different classes called  Disks, FreeDisks, UsedDisks in persistence
+  module.
+* Modify the manager to retrieve inventory details from tendrl-node-inventory
+  file and compare it with node_inventory list, If any changes then populate the
+  disk details in class instance and store it in etcd.
+```
+- Populate all disk details in Disks class instance and store it in etcd.
+- Populate the free disks id in FreeDisks class instance and store it in etcd.
+- Populate the used disks id in UsedDisks class instance and store it in etcd.
+```
+* Store all three class instances in etcd with different keys.
+* Create objects in tendrl definition for disks, used disks and free disks.
+* Disks structure in etcd:
+```
+/nodes/node_id/Disks
+   <disk_id1> # object with all the details for disk_id1
+   <disk_id2> # object with all the details for disk_id2
+
+/nodes/node_id/Disks/used
+   <disk_id1> # empty object
+
+/nodes/node_id/Disks/free
+   <disk_id2> #empty object
+```
+
+== Alternatives
+
+* hdparm is also giving some disk details like serial number and firmware but it
+  works only for ATA disks types. For SCSI disk we have to use sdparm.
+
+== Data model impact
+
+* Add used disks id in etcd with key nodes/node_id/Disks/Used.
+* Add free disks id in etcd with key nodes/node_id/Disks/Free.
+* Add all disks details in etcd with key nodes/node_id/Disks
+* Add three different object in tendrl definitions for disks, used disks and
```

Comment:
Not three different objects- one object per disk. I'd mentioned this in my previous comment yesterday, but it seems to have been lost.

There needs to be one object per disk, which contains all the information about the disk. This needs to be in the `disks` directory. Lower case, not capitalised. The object's key needs to be the id that's guaranteed to always be available.

The objects under `used` and `free` directories under `disks` (again, lower case, not capitalised), need to be empty objects with their key being the id of the disk object under `disks`.

Add validations to ensure that the same id doesn't fall under both `used` and `free`.